### PR TITLE
[F] Metis: Restructure result that is returned

### PR DESCRIPTION
**Pull request by DrWilliamsSteven**
For mapping required by @mareq indexing,

This PR is to:
- [x] restructure the result returned, so that an array of objects is returned.
- [x] returns a numerical confidence score

Relevant to repositive/agora#139

TO DO IN A FUTURE PR:
- [ ] Change so that ontology is not merged into dataset
- [ ] store results in separate postgres table rather than merging into dataset

### CODE REVIEW

File: src/annotate.ts [hunk 1]
```diff
@@ -7,10 +7,21 @@ import * as fs from 'fs';
...
+  const obj = {
+    'LOW': 0,
+    'MEDIUM': 0,
+    'GOOD': 0,
+    'HIGH': 0
+  };
+  obj[string] = 1;
+  return obj;
+}
+
```
Comment (DrWilliamsSteven) [inline src/annotate.ts]: Is there a better way to write this?

File: src/annotate.ts [hunk 2]
```diff
@@ -43,43 +54,48 @@ export default async function annotate({payload, _ajv = ajv, _schema = schema}:
...
-  json = json[0];
-  //console.log(JSON.stringify(json, undefined, 2));
-  // if there is no match, returns undefined
-  // if there are multiple matches, returns only the first one.. check this is appropriate.
+  // if there is no match, returns [undefined]
+  // if there are multiple matches, returns all.
 
-  if (json) {
+  if (json.length > 0 && json[0] !== undefined) {
+    return json.map((x: any) => {
```
Comment (DrWilliamsSteven) [inline src/annotate.ts]: Is there a better way to check whether 
```
json !== [undefined]
```

**Comment by DrWilliamsSteven** [comment]
Also while I think of it, can you please add 'Review is required to merge PR' to this repo? Thanks.

File: src/annotate.ts [hunk 3]
```diff
@@ -7,10 +7,21 @@ import * as fs from 'fs';
 const schema: any = JSON.parse(fs.readFileSync('./schemas/annotate-is-valid.json', 'utf8'));
 const ajv = new Ajv({ allErrors: true, verbose: true });
 
-export default async function annotate({payload, _ajv = ajv, _schema = schema}: { payload: { field: string, term: string }, _ajv: typeof ajv, _schema: typeof schema}) {
+function confidenceScoreF(string: string) {
```
Comment (Istar-Eldritch) [inline src/annotate.ts]: Could you add a comment explaining what you try to do with this function?
Do not use the word `string` as the name of argument, it clashes with the reserved word for the type, use something like `str` instead.

File: src/annotate.ts [hunk 4]
```diff
@@ -7,10 +7,21 @@ import * as fs from 'fs';
...
+  const obj = {
+    'LOW': 0,
+    'MEDIUM': 0,
+    'GOOD': 0,
+    'HIGH': 0
+  };
+  obj[string] = 1;
```
Comment (Istar-Eldritch) [inline src/annotate.ts]: I don't know if you really want to do this, but if so, do object assignation instead:
```ts
return {...obj, [key]: 1}
```

File: src/annotate.ts (hunk 2 above)
Comment (Istar-Eldritch) [reply src/annotate.ts]: Why do you need to check for  the first entry in the array? `json[0] !== undefined`. I would check if it's an array with [Array.isArray](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/isArray), you can skyp the null & undefined checks as the function will do it for you.

File: src/annotate.ts [hunk 5]
```diff
@@ -43,43 +54,48 @@ export default async function annotate({payload, _ajv = ajv, _schema = schema}:
...
 
-    const result = {
-      originalTerm: term,
-      ontologyTerm: json.annotatedProperty.propertyValue.toLowerCase(),
-      ontologyIRI: json._links.olslinks[0].semanticTag,
-      ontologyConfidence: json.confidence,
+      return {
+        term: x.annotatedProperty.propertyValue.toLowerCase(),
+        iri: x._links.olslinks[0].semanticTag,
+        confidence: confidenceScoreF(x.confidence),
```
Comment (Istar-Eldritch) [inline src/annotate.ts]: Is this property going to be an object or a single scalar value?

File: src/annotate.ts (hunk 2 above)
Comment (DrWilliamsSteven) [reply src/annotate.ts]: The request returns `[undefined]` if the term isn't found. So checking it is an array isn't enough.

File: src/annotate.ts (hunk 5 above)
Comment (DrWilliamsSteven) [reply src/annotate.ts]: @mareq asked for the object structure I've returned from the confidenceScoreF function

File: src/annotate.ts [hunk 6]
```diff
@@ -43,43 +43,38 @@ export default async function annotate({payload, _ajv = ajv, _schema = schema}:
...
 
-  json = json[0];
-  //console.log(JSON.stringify(json, undefined, 2));
-  // if there is no match, returns undefined
-  // if there are multiple matches, returns only the first one.. check this is appropriate.
+  // if there is no match, the request returns [undefined]
+  // if there are multiple matches, returns all.
 
-  if (json) {
+  if (json.length > 0 && typeof json[0] !== 'undefined') {
```
Comment (DrWilliamsSteven) [inline src/annotate.ts]: @Istar-Eldritch I think this way checks correctly what I need.
if term = 'cancer', then json returned from request = [{results: etc} ] 
but
if term = 'zzzz', then json returned from request = [undefined]

Hence checking that json[0]  isn't undefined.

Comment (Istar-Eldritch) [reply src/annotate.ts]: The array has a string (`'undefined'`) as the first position?

Comment (DrWilliamsSteven) [reply src/annotate.ts]: No when the term isn't found it returns 
```
[ undefined ] 
```

Comment (Istar-Eldritch) [reply src/annotate.ts]: Oh yeah, then you can just do `if (json[0])` but do it only if you are sure json is an array.

File: src/annotate.ts [hunk 7]
```diff
@@ -43,43 +43,38 @@ export default async function annotate({payload, _ajv = ajv, _schema = schema}:
...
+      if (short_name === 'orphanet') short_name = 'ordo';
 
-    // make ontologyShortName equal to ontology accession prefix
-    let ontologyShortName = uriSplit.slice(-1)[0].split('_')[0].toLowerCase();
-    if (['format', 'operation', 'topic'].indexOf(ontologyShortName) > -1) ontologyShortName = 'edam';
-    if (ontologyShortName === 'orphanet') ontologyShortName = 'ordo';
+      return {
+        term: x.annotatedProperty.propertyValue.toLowerCase(),
+        iri: x._links.olslinks[0].semanticTag,
+        confidence: x.confidence,
```
Comment (Istar-Eldritch) [inline src/annotate.ts]: We are still returning confidence as a `string`. Is this the expected behaviour for this PR?

**Comment by Istar-Eldritch** [comment]
Sorry for taking that much time to review this.

As you return an array when the behaviour is correct I modified slightly the code so it will always return an array, even if there are no matches, in which case it will return an empty array. This should be easier to handle in clients than having to deal with the undefined.

I also added a bit of docs on what you should expect as a response in the README.

**Pull request merged**